lib/iris/cube.py

```diff
@@ -1642,11 +1671,23 @@ def lazy_data(self, array=None):
                 if self.shape or array.shape != (1,):
                     raise ValueError('Require cube data with shape %r, got '
                                      '%r.' % (self.shape, array.shape))
-            self._my_data = array
-        else:
-            array = self._my_data
-        array = as_lazy_data(array)
-        return array
+            self._dask_array = array
+            self._numpy_array = None
+            result = self._dask_array
+        elif self._numpy_array is not None:
+            data = self._numpy_array
+            if isinstance(data, np.ma.masked_array):
+                if np.ma.is_masked(data):
+                    if data.dtype.kind == 'i':
+                        data = data.astype(np.dtype('f8'))
+                    # Where possible, write these NANs into the cube's
+                    # _numpy_array.
+                    data.data[data.mask] = np.nan
+                data = data.data
```

Comment:
@marqh This block of code is identical to [this](https://github.com/SciTools/iris/pull/2392/files#diff-4663903670147de0963e3cd8b176d753R1233) ... we need to refactor this out.